contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -98,29 +85,34 @@ contract MixinExchangeWrapper is
             uint256 makerAssetAcquiredAmount
         )
     {
-        // No fee or percentage fee
+        uint256 protocolFee = tx.gasprice.safeMul(EXCHANGE.protocolFeeMultiplier());
+
+        // No taker fee or percentage fee
         if (order.takerFee == 0 || order.takerFeeAssetData.equals(order.makerAssetData)) {
             // Attempt to sell the remaining amount of WETH
             LibFillResults.FillResults memory singleFillResults = _fillOrderNoThrow(
                 order,
-                remainingTakerAssetFillAmount,
+                remainingTakerAssetFillAmount.safeSub(protocolFee),
                 signature
             );
 
-            wethSpentAmount = singleFillResults.takerAssetFilledAmount;
+            wethSpentAmount = singleFillResults.takerAssetFilledAmount.safeAdd(
+                singleFillResults.protocolFeePaid
+            );
 
             // Subtract fee from makerAssetFilledAmount for the net amount acquired.
             makerAssetAcquiredAmount = singleFillResults.makerAssetFilledAmount.safeSub(
                 singleFillResults.takerFeePaid
             );
         // WETH fee
         } else if (order.takerFeeAssetData.equals(order.takerAssetData)) {
+
             // We will first sell WETH as the takerAsset, then use it to pay the takerFee.
-            // This ensures that we reserve enough to pay the fee.
+            // This ensures that we reserve enough to pay the taker and protocol fees.
             uint256 takerAssetFillAmount = LibMath.getPartialAmountCeil(
                 order.takerAssetAmount,
                 order.takerAssetAmount.safeAdd(order.takerFee),
-                remainingTakerAssetFillAmount
+                remainingTakerAssetFillAmount.safeSub(protocolFee)
```

Comment:
I'm not sure we actually need to protect the user from not having enough to pay the protocol fee. If they did send in enough ETH to cover it with the transaction, then this logic ends up under filling the orders.

Reply:
Sorry for the thrash on this one, but I think it would actually be cleaner to just do this calculation before calling `_marketSellSingleOrder`, rather than passing in the `protocolFee`. `remainingTakerAssetFillAmount` is always used in this way, so might as well combine them.